src/libsystemd-network/radv-internal.h

```diff
@@ -63,19 +63,34 @@ struct sd_radv {
         struct sd_radv_opt_dns *dnssl;
 };
 
+#define radv_prefix_opt__contents {             \
+        uint8_t type;                           \
+        uint8_t length;                         \
+        uint8_t prefixlen;                      \
+        uint8_t flags;                          \
+        be32_t valid_lifetime;                  \
+        be32_t preferred_lifetime;              \
+        uint32_t reserved;                      \
+        struct in6_addr in6_addr;               \
+}
+
+struct radv_prefix_opt radv_prefix_opt__contents;
+
+/* We need the opt substructure to be packed, because we use it in send(). But
+ * if we use _packed_, this means that the structure cannot be used directly in
+ * normal code in general, because the fields might not be properly aligned.
+ * But in this particular case, the structure is defined in a way that gives
+ * proper alignment, even without the explicit _packed_ attribute. To appease
+ * the compiler we use the "unpacked" structure, but we also verify that
+ * structure contains no holes, so offsets are the same when _packed_ is used.
+ */
+struct _packed_ radv_prefix_opt__packed radv_prefix_opt__contents;
+assert_cc(sizeof(struct radv_prefix_opt) == sizeof(struct radv_prefix_opt__packed));
```

Comment:
i wonder if we could somehow do a check that also checks the inner alignments are the same... maybe at least add:

```c
assert_cc(offsetof(struct radv_prefix_opt, in6_addr) == offsetof(struct radv_prefix_op__packet, in6_addr));
```

(because that one field happens to be at the end of everything, and has biggest alignment)

Reply:
btw, given that this structure just maps 1:1 to bitwise definitions in the RFC I think doing an assert_cc() that just compares the resulting struct size (and maybe some offsets) to the literal byte values from the RFC would be OK too, i.e. without the whole _packed_ double definition

Reply:
> do a check that also checks the inner alignments are the same

I started with that, but actually, it is redundant, IIUC. When the structure is packed, there can be no padding internally or at the end. So if the size of the "unpacked" structure is the same, it means it has no padding either.

> given that this structure just maps 1:1 to bitwise definitions in the RFC I think doing an assert_cc() that just compares the resulting struct size (and maybe some offsets) to the literal byte values from the RFC would be OK too

I guess. But since this PR is already written, I'd prefer to just keep it. I don't feel like going through the RFC and copying the offsets and whatnot.